File: surface/widgets/swiftui/components/widgets/concepts/governance-decision/VoteResultBar.swift
import SwiftUI

struct VoteSegment: Identifiable {
    let id = UUID()
    let label: String; let count: Int; var color: Color?
}

enum VoteResultBarWidgetState { case idle, animating, segmentHovered }

struct VoteResultBarView: View {
    let segments: [VoteSegment]
    var total: Int?
    var variant: String = "binary"
    var showLabels: Bool = true
    var showQuorum: Bool = false
    var quorumThreshold: Double = 0
    var animate: Bool = true
    var size: String = "md"
    var onSegmentHover: ((Int?, VoteSegment?) -> Void)?

    @State private var widgetState: VoteResultBarWidgetState = .idle
    @State private var hoveredIndex: Int?

    private let defaultColors: [Color] = [.green, .red, .orange, .blue, .purple, .cyan]
    private var effectiveTotal: Int { total ?? segments.reduce(0) { $0 + $1.count } }
    private var barHeight: CGFloat { size == "sm" ? 16 : size == "lg" ? 36 : 24 }

    private func percent(_ count: Int) -> Double {
        effectiveTotal > 0 ? Double(count) / Double(effectiveTotal) * 100 : 0
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 4) {
            // Bar
            GeometryReader { geo in
                ZStack(alignment: .leading) {
                    HStack(spacing: 0) {
                        ForEach(Array(segments.enumerated()), id: \.element.id) { idx, seg in
                            let w = geo.size.width * CGFloat(percent(seg.count) / 100)
                            Rectangle()
                                .fill(seg.color ?? defaultColors[idx % defaultColors.count])
                                .frame(width: max(w, seg.count == 0 ? 2 : w))
                                .opacity(hoveredIndex != nil && hoveredIndex != idx ? 0.5 : 1)
                                .onHover { h in hoveredIndex = h ? idx : nil; onSegmentHover?(h ? idx : nil, h ? seg : nil) }
                        }
                    }
                    if showQuorum && quorumThreshold > 0 {
                        Rectangle().fill(Color.black).frame(width: 2)
                            .offset(x: geo.size.width * CGFloat(quorumThreshold / 100))
                    }
                }
            }
            .frame(height: barHeight)
            .cornerRadius(4)

            if showLabels {
                HStack(spacing: 12) {
                    ForEach(Array(segments.enumerated()), id: \.element.id) { idx, seg in
                        HStack(spacing: 4) {
                            Circle().fill(seg.color ?? defaultColors[idx % defaultColors.count]).frame(width: 8, height: 8)
                            Text("\(seg.label) \(seg.count) (\(String(format: "%.1f", percent(seg.count)))%)")
                                .font(.system(size: size == "sm" ? 11 : size == "lg" ? 14 : 12))
                                .foregroundColor(.secondary)
                        }
                    }
                }
            }

            Text("Total: \(effectiveTotal)").font(.system(size: size == "sm" ? 11 : 12)).foregroundColor(.secondary)
        }
        .accessibilityElement(children: .contain)
        .accessibilityLabel("Vote results, total \(effectiveTotal) votes")
    }
}

File: surface/widgets/compose/components/widgets/concepts/governance-decision/VoteResultBar.kt
package com.clef.surface.widgets.concepts.governancedecision

import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import kotlinx.coroutines.delay

// --- State machine ---

enum class VoteResultBarState { Idle, Animating, SegmentHovered }

sealed class VoteResultBarEvent {
    data class HoverSegment(val index: Int) : VoteResultBarEvent()
    object AnimateIn : VoteResultBarEvent()
    object AnimationEnd : VoteResultBarEvent()
    object Unhover : VoteResultBarEvent()
}

fun voteResultBarReduce(state: VoteResultBarState, event: VoteResultBarEvent): VoteResultBarState = when (state) {
    VoteResultBarState.Idle -> when (event) {
        is VoteResultBarEvent.HoverSegment -> VoteResultBarState.SegmentHovered
        is VoteResultBarEvent.AnimateIn -> VoteResultBarState.Animating
        else -> state
    }
    VoteResultBarState.Animating -> when (event) {
        is VoteResultBarEvent.AnimationEnd -> VoteResultBarState.Idle
        else -> state
    }
    VoteResultBarState.SegmentHovered -> when (event) {
        is VoteResultBarEvent.Unhover -> VoteResultBarState.Idle
        is VoteResultBarEvent.HoverSegment -> VoteResultBarState.SegmentHovered
        else -> state
    }
}

// --- Types ---

data class VoteSegment(
    val label: String,
    val count: Int,
    val color: Color? = null
)

private val DEFAULT_COLORS = listOf(
    Color(0xFF4CAF50), Color(0xFFF44336), Color(0xFFFF9800), Color(0xFF2196F3),
    Color(0xFF9C27B0), Color(0xFF00BCD4), Color(0xFF795548), Color(0xFF607D8B)
)

private fun formatPercent(value: Float): String {
    val formatted = "%.1f".format(value)
    return if (formatted.endsWith(".0")) formatted.dropLast(2) else formatted
}

@Composable
fun VoteResultBar(
    segments: List<VoteSegment>,
    modifier: Modifier = Modifier,
    total: Int? = null,
    variant: String = "binary",
    showLabels: Boolean = true,
    showQuorum: Boolean = false,
    quorumThreshold: Float = 0f,
    animate: Boolean = true,
    size: String = "md",
    onSegmentHover: (Int?, VoteSegment?) -> Unit = { _, _ -> }
) {
    var state by remember { mutableStateOf(VoteResultBarState.Idle) }
    var hoveredIndex by remember { mutableStateOf<Int?>(null) }
    var animated by remember { mutableStateOf(!animate) }

    val resolvedTotal = remember(total, segments) {
        if (total != null && total > 0) total else segments.sumOf { it.count }
    }

    data class ComputedSegment(
        val segment: VoteSegment,
        val percent: Float,
        val resolvedColor: Color
    )

    val computedSegments = remember(segments, resolvedTotal) {
        segments.mapIndexed { i, seg ->
            val percent = if (resolvedTotal > 0) (seg.count.toFloat() / resolvedTotal) * 100f else 0f
            ComputedSegment(seg, percent, seg.color ?: DEFAULT_COLORS[i % DEFAULT_COLORS.size])
        }
    }

    val ariaDescription = remember(computedSegments, resolvedTotal) {
        val parts = computedSegments.joinToString(", ") {
            "${it.segment.label}: ${it.segment.count} votes (${formatPercent(it.percent)}%)"
        }
        "Vote results: $parts. Total: $resolvedTotal votes."
    }

    // Animation on mount
    LaunchedEffect(animate) {
        if (animate) {
            state = voteResultBarReduce(state, VoteResultBarEvent.AnimateIn)
            delay(50) // let compose paint zero widths first
            animated = true
            delay(400)
            state = voteResultBarReduce(state, VoteResultBarEvent.AnimationEnd)
        } else {
            animated = true
        }
    }

    val barHeight = when (size) {
        "sm" -> 16.dp
        "lg" -> 36.dp
        else -> 24.dp
    }
    val labelFontSize = when (size) {
        "sm" -> 11.sp
        "lg" -> 14.sp
        else -> 12.sp
    }

    Column(modifier = modifier.semantics { contentDescription = ariaDescription }) {
        // Bar
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .height(barHeight)
                .clip(MaterialTheme.shapes.small)
                .background(Color(0xFFE0E0E0))
        ) {
            Row(modifier = Modifier.fillMaxSize()) {
                computedSegments.forEachIndexed { i, seg ->
                    val targetWeight = if (animated) seg.percent.coerceAtLeast(0.1f) else 0.1f
                    val animatedWeight by animateFloatAsState(
                        targetValue = targetWeight,
                        animationSpec = tween(if (animate) 400 else 0),
                        label = "segment-$i"
                    )
                    val isHovered = hoveredIndex == i
                    val alpha = if (hoveredIndex != null && !isHovered) 0.5f else 1f

                    Box(
                        modifier = Modifier
                            .weight(animatedWeight)
                            .fillMaxHeight()
                            .background(seg.resolvedColor.copy(alpha = alpha))
                    )
                }
            }

            // Quorum marker
            if (showQuorum && quorumThreshold > 0f) {
                Box(
                    modifier = Modifier
                        .fillMaxHeight()
                        .width(2.dp)
                        .offset(x = (quorumThreshold / 100f * 300).dp) // approximate
                        .background(Color.Black)
                )
            }
        }

        // Labels
        if (showLabels) {
            Row(
                horizontalArrangement = Arrangement.spacedBy(12.dp),
                modifier = Modifier.padding(top = 4.dp).fillMaxWidth()
            ) {
                computedSegments.forEach { seg ->
                    Row(
                        verticalAlignment = Alignment.CenterVertically,
                        horizontalArrangement = Arrangement.spacedBy(4.dp)
                    ) {
                        Box(
                            modifier = Modifier
                                .size(8.dp)
                                .clip(CircleShape)
                                .background(seg.resolvedColor)
                        )
                        Text(
                            "${seg.segment.label} ${seg.segment.count} (${formatPercent(seg.percent)}%)",
                            fontSize = labelFontSize,
                            color = MaterialTheme.colorScheme.onSurfaceVariant
                        )
                    }
                }
            }
        }

        // Total
        Text(
            "Total: $resolvedTotal",
            fontSize = labelFontSize,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
            modifier = Modifier.padding(top = 4.dp)
        )
    }
}
